src/resolve.ts

```diff
@@ -0,0 +1,47 @@
+import { createRequire as req } from 'node:module'
+import { expiringMap } from './expiring-map'
+
+const localRequire = req(import.meta.url)
+
+// This is a long-lived cache for resolved modules whether they exist or not
+// Because we're compatible with a large number of plugins, we need to check
+// for the existence of a module before attempting to import it. This cache
+// is used to mitigate the cost of that check because Node.js does not cache
+// failed module resolutions making repeated checks very expensive.
+const resolveCache = expiringMap<string, string | null>(30_000)
+
+export function resolveIn(id: string, paths: string[]) {
+  return localRequire.resolve(id, {
+    paths,
+  })
+}
+
+export function maybeResolve(name: string) {
+  let modpath = resolveCache.get(name)
+
+  if (modpath === undefined) {
+    modpath = freshMaybeResolve(name)
+    resolveCache.set(name, modpath)
```

Comment:
Is the idea that if this still fails, we will fill the cache with `null` instead of the default `undefined` marking it as "we tried, but didn't work"?

Reply:
Correct!